cmd/pfaFindFast/pfaFindFast.go

```diff
@@ -0,0 +1,147 @@
+// Command Group: "Sequence Evolution & Reconstruction"
+
+// For two pFasta sequences, calculate a score at each base for the divergence between the two pDNA bases, convert the score to a binary substitution/non-substitution call, and return the calculations over a given window size
+package main
+
+import (
+	"flag"
+	"fmt"
+	"github.com/vertgenlab/gonomics/dna/pDna"
+	"github.com/vertgenlab/gonomics/fasta/pFasta"
+	"log"
+	"math"
+	"os"
+	"runtime/pprof"
+)
+
+type Settings struct {
+	InFile                  string
+	OutFile                 string
+	FirstQueryName          string
+	SecondQueryName         string
+	WindowSize              int
+	RefChromName            string
+	RemoveN                 bool
+	LongOutput              bool
+	DivergenceRate          float64
+	OutputAlnPos            bool
+	BaseDistToDivThreshold  float64
+	BaseDotToSubstThreshold float64
+}
+
+func pfaFindFast(s Settings) {
+	var reference, firstQuery, secondQuery []pDna.Float32Base
+	var found bool
+
+	records := pFasta.Read(s.InFile)    // fasta.Read already makes sure that sequence names are unique
+	recordsMap := pFasta.ToMap(records) // generate map[string][]dna.Base. No order but can do key-value search
+
+	if len(records) < 2 {
+		log.Fatalf("Error: There must be at least 2 pFasta records in the input file.\n")
+	}
+
+	//if reference and query names were not specified in the command, then use 1st and 2nd fasta records' names, respectively
+	//allows backward compatibility with previously-written code
+	if s.FirstQueryName == "" {
+		firstQuery = records[0].Seq
+	} else {
+		_, found = recordsMap[s.FirstQueryName]
+		if found {
+			firstQuery = recordsMap[s.FirstQueryName]
+		} else {
+			log.Fatalf("Error: first query name is not found in the input file.\n")
+		}
+	}
+	if s.SecondQueryName == "" {
+		secondQuery = records[1].Seq
+	} else {
+		_, found = recordsMap[s.SecondQueryName]
+		if found {
+			secondQuery = recordsMap[s.SecondQueryName]
+		} else {
+			log.Fatalf("Error: second query name is not found in the input file.\n")
+		}
+	}
+
+	reference = records[0].Seq // reference will always be the 1st pFasta record in a multi-pFa input file
+
+	if !(len(reference) == len(firstQuery) && len(reference) == len(secondQuery)) {
+		log.Fatalf("Error: Reference, first query, and second query sequences are not all of equal length.\n")
+	}
+
+	speedyWindowDifference(reference, firstQuery, secondQuery, s)
+}
+
+func usage() {
+	fmt.Print(
+		"pfaFindFast - For two pFasta sequences, calculate scores at each base for the divergence that separates the two pDNA bases, and return a bed file with calculations.\n" +
```

Comment:
Example script:
support.doFastestNeutralRate.d.csh
#!/bin/csh -evf

set reconstructSeq = /hpc/group/vertgenlab/raven/GOPATH/src/github.com/vertgenlab/gonomics/cmd/reconstructSeq/reconstructSeq
set faFindFast = /hpc/group/vertgenlab/raven/GOPATH/src/github.com/vertgenlab/gonomics/cmd/pfaFindFast/pfaFindFast
set faFilter = ~/go/bin/faFilter
set bedMerge = ~/go/bin/bedMerge

set mod = /hpc/group/vertgenlab/publicDataSets/PrimateT2T_15way/pfaHaqer/all-4d.mod
set inputFaDir = /hpc/group/vertgenlab/publicDataSets/PrimateT2T_15way/multizFa
set tmp = /work/yl726/PrimateT2T_15way
set haqerTmp = $tmp/haqerTmp
set outDir = $tmp/fastest10mbWindows

set chr = "$1"
set windowSize = 10_000_000

echo $chr

$reconstructSeq -pDnaOutFile $haqerTmp/$chr.pFa -pDnaNodeMulti=hca,HUMANanc -keepAllSeq $mod $inputFaDir/$chr.fa $haqerTmp/$chr.mleRecon.forpfa.fa.gz

$faFindFast -longOutput -chrom $chr -windowSize $windowSize -firstQueryName HUMANanc -secondQueryName hca -outputAlnPos $haqerTmp/$chr.pFa $haqerTmp/$chr.mleRecon.10mbDiff.bed.gz
zcat $haqerTmp/$chr.mleRecon.10mbDiff.bed.gz | $bedMerge -lowMem -mergeAdjacent stdin $outDir/$chr.HighestScore.txt

echo DONE

Reply:
Output fields:
1.s.RefChromName, 2.refIdxWindowStart, 3.lastRefIdxOfWindowPlusOne, 4.s.RefChromName_refIdxWindowStart, 
5.windowDotSubst+totalGaps, 
6."+", 7.percentDiverged, 8.rawPValue, 9.alnIdxBeforeWindow+1, 
10.windowDotSubst, 11.windowDotMean, 12.windowDot (whole-window field, often cut out), 
13.windowDistDiv, 14.windowDistMean, 15.windowDist (whole-window field, often cut out)

Reply:
(base) yl726@dcc-login-02  /work/yl726/PrimateT2T_15way/haqerTmp $ zcat chr21.mleRecon.10mbDiff.bed.gz | head | cut -f 1-11,13-14
chr21	0	14883625	chr21_0	68569	+	6.856900e-01	0.000000e+00	0	56666	3.866303e-01	122972	4.314101e-02
chr21	3291	14883626	chr21_3291	68569	+	6.856900e-01	0.000000e+00	3390	56666	3.864945e-01	122972	4.315056e-02
chr21	3292	14883627	chr21_3292	68569	+	6.856900e-01	0.000000e+00	3391	56666	3.864945e-01	122972	4.315056e-02
chr21	3293	14883628	chr21_3293	68569	+	6.856900e-01	0.000000e+00	3392	56666	3.864945e-01	122972	4.315056e-02
chr21	3294	14883629	chr21_3294	68569	+	6.856900e-01	0.000000e+00	3393	56666	3.864945e-01	122972	4.315056e-02
chr21	3295	14883630	chr21_3295	68569	+	6.856900e-01	0.000000e+00	3394	56666	3.864945e-01	122972	4.315056e-02
chr21	3296	14883631	chr21_3296	68569	+	6.856900e-01	0.000000e+00	3395	56666	3.864945e-01	122972	4.315056e-02
chr21	3297	14883632	chr21_3297	68569	+	6.856900e-01	0.000000e+00	3396	56666	3.864945e-01	122972	4.315056e-02
chr21	3298	14883633	chr21_3298	68569	+	6.856900e-01	0.000000e+00	3397	56666	3.864945e-01	122972	4.315056e-02
chr21	3299	14883634	chr21_3299	68569	+	6.856900e-01	0.000000e+00	3398	56666	3.864945e-01	122972	4.315056e-02

Reply:
Ways to improve efficiency:
1. Reduce write-to-file fields
2. Make substitution calculation more efficient, in efficient.go incrementWindowEdge function as opposed to right before writing out window to bed
3. Have flags, e.g. -longOutput for whole-window output, option for either dot product or Euclidean not both